File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/move/Move.java
package org.akoshterek.backgammon.move;

import org.akoshterek.backgammon.eval.Reward;
import org.akoshterek.backgammon.board.PositionClass;

import java.util.Comparator;

/**
 * @author Alex
 *         date 20.07.2015.
 */
public class Move {
    public ChequersMove anMove = new ChequersMove();
    public AuchKey auch = new AuchKey();
    public int cMoves, cPips;
    /* scores for this move */
    public double rScore;
    /* evaluation for this move */
    public Reward arEvalMove = new Reward();
    public PositionClass pc;
    public int backChequer;

    private static int compareMoves(final Move pm0, final Move pm1) {
        if(pm1.rScore ==  pm0.rScore) return 0;
		/*high score first */
        return (pm1.rScore > pm0.rScore) ? 1 : -1;
    }

    public static final Comparator<Move> moveComparator = new MoveComparator();
    // compare moves according to equity and back chequer
    private static class MoveComparator implements Comparator<Move> {
        public int compare(final Move p, final Move q) {
            if(p == q) return 0;

            if (p.rScore != q.rScore)
                return compareMoves(p, q);

            if(p.backChequer == q.backChequer) {
                return 0;
            } else {
                return (p.backChequer > q.backChequer) ? 1 : -1;
            }
        }
    }
}

File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/move/Move.scala
package org.akoshterek.backgammon.move

import org.akoshterek.backgammon.eval.Reward
import org.akoshterek.backgammon.board.PositionClass

/**
  * @author Alex
  *         date 20.07.2015.
  */

object Move extends Ordering[Move] {
  def compare(x: Move, y: Move): Int = {
    if (x eq y) {
      0
    } else if (x.rScore != y.rScore) {
      // smaller score is 'smaller' move
      x.rScore.compareTo(y.rScore)
    } else {
      // bigger back chequer is 'smaller' move
      y.backChequer.compareTo(x.backChequer)
    }
  }
}

class Move {
  var anMove: ChequersMove = new ChequersMove
  var auch: AuchKey = new AuchKey
  var cMoves = 0
  var cPips = 0
  /** scores for this move */
  def rScore = arEvalMove.equity
  /** evaluation for this move */
  var arEvalMove: Reward = new Reward()
  var pc: PositionClass = PositionClass.CLASS_OVER
  var backChequer = 0
}